### PR TITLE
LG-2923: Updates useIdAllocator

### PR DESCRIPTION
**Pull request by bruugey**
## ✍️ Proposed changes

Temp solution for useIdAllocator, to improve parity between client-side and SSR

🎟 _Jira ticket:_ [LG-2923](https://jira.mongodb.org/browse/LG-2923)

### For bug fixes, new features & breaking changes

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have run `yarn changeset` and documented my changes

### For new components

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README

### CODE REVIEW

File: packages/hooks/src/useIdAllocator.ts [hunk 1]
```diff
@@ -1,53 +1,44 @@
...
+    idOverride,
+  );
 
   useEffect(() => {
-    if (serverHandoffComplete === false) {
-      serverHandoffComplete = true;
+    if (defaultId == null) {
+      // Fallback to this default id when possible.
+      // Use the incrementing value for client-side rendering only.
+      // We can't use it server-side.
```
Comment (shaneeza) [inline packages/hooks/src/useIdAllocator.ts:21]: We just spoke about this, but I'm confused about this comment. If we're not incrementing the value server-side, then what is returned server-side? it looks like `undefined` if there is no `idOverride`?

File: packages/hooks/src/useIdAllocator.ts [hunk 2]
```diff
@@ -1,53 +1,44 @@
...
+];
+
+export default function useId({ prefix, id: idOverride }: Params): string {
+  if (maybeReactUseId !== undefined) {
+    const reactId = maybeReactUseId();
+    return idOverride ?? reactId;
+  }
+
+  // eslint-disable-next-line react-hooks/rules-of-hooks -- `React.useId` is invariant at runtime.
+  return useGlobalId({ id: idOverride, prefix });
```
Comment (TheSonOfThomp) [inline packages/hooks/src/useIdAllocator.ts:43]: We should probably still follow the rules of hooks. 
```tsx
const globalId = useGlobalId(...)
if (maybeReactUseId) {
  return ...
}
return globalId
```

File: packages/hooks/src/useIdAllocator.ts [hunk 3]
```diff
@@ -1,53 +1,44 @@
...
- * @param options.prefix string that prefixes the generated id
- * @param options.id string that represents an already-created id
- */
-function useIdAllocator({
-  prefix,
-  id: idProp,
-}: {
+// Currently using Material UI useId hook until we can upgrade to React 18's useId
+// https://github.com/mui/material-ui/blob/master/packages/mui-utils/src/useId.ts
+import React, { useEffect, useState } from 'react';
```
Comment (TheSonOfThomp) [inline packages/hooks/src/useIdAllocator.ts:3]: nit: can we move this to its own `useIdAllocator/` directory, with its own `useIdAllocator.spec.ts` file ?

File: .changeset/fresh-spies-act.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+---
+'@leafygreen-ui/a11y': patch
+'@leafygreen-ui/checkbox': patch
```
[the remark this answers is not in the document]
Comment (bruugey) [reply .changeset/fresh-spies-act.md:3]: Did you run yarn clean && yarn run init
I am not seeing these issues anymore locally

[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply .changeset/fresh-spies-act.md:3]: Ah, yeah. Not seeing the errors after that

**Review by TheSonOfThomp — APPROVED** [review]

**Pull request merged**